src/Metadata/ToMetadata.purs

```diff
@@ -0,0 +1,85 @@
+module Metadata.ToMetadata
+  ( class ToMetadata
+  , toMetadata
+  , AnyToMetadata
+  , class AnyToMetadataClass
+  , anyToMetadata
+  ) where
+
+import Prelude
+
+import Data.Array (fromFoldable) as Array
+import Data.BigInt (BigInt)
+import Data.Foldable (class Foldable)
+import Data.Map (Map)
+import Data.Map (catMaybes, fromFoldable, toUnfoldable) as Map
+import Data.Maybe (Maybe(Just), fromJust)
+import Data.NonEmpty (NonEmpty)
+import Data.Profunctor.Strong ((***))
+import Data.Tuple.Nested (type (/\))
+import Partial.Unsafe (unsafePartial)
+import Types.ByteArray (ByteArray)
+import Types.Int (Int, fromBigInt) as Int
+import Types.TransactionMetadata
+  ( TransactionMetadatum(MetadataMap, MetadataList, Int, Bytes, Text)
+  )
+
+--------------------------------------------------------------------------------
+-- ToMetadata
+--------------------------------------------------------------------------------
+
+class ToMetadata (a :: Type) where
+  toMetadata :: a -> TransactionMetadatum
+
+instance ToMetadata TransactionMetadatum where
+  toMetadata = identity
+
+instance (Ord k, ToMetadata k) => ToMetadata (Map k AnyToMetadata) where
+  toMetadata =
+    toMetadata <<< Map.catMaybes <<< map (\(AnyToMetadata f) -> f toMetadata)
+else instance (ToMetadata k, ToMetadata v) => ToMetadata (Map k v) where
+  toMetadata mp =
+    let
+      entries = Map.toUnfoldable mp :: Array (k /\ v)
+    in
+      MetadataMap <<< Map.fromFoldable $
+        map (toMetadata *** toMetadata) entries
+
+instance (Ord k, ToMetadata k) => ToMetadata (Array (k /\ AnyToMetadata)) where
+  toMetadata = toMetadata <<< Map.fromFoldable
+else instance (Ord k, ToMetadata k, ToMetadata v) => ToMetadata (Array (k /\ v)) where
+  toMetadata = toMetadata <<< Map.fromFoldable
+else instance ToMetadata a => ToMetadata (Array a) where
+  toMetadata = MetadataList <<< map toMetadata
+
+instance (Foldable f, ToMetadata a) => ToMetadata (NonEmpty f a) where
+  toMetadata = toMetadata <<< Array.fromFoldable
+
+instance ToMetadata Int.Int where
+  toMetadata = Int
+
+-- FIXME: Come up with a type-safe error handling approach.
+instance ToMetadata BigInt where
+  toMetadata bi =
+    unsafePartial $ Int $ fromJust $ Int.fromBigInt bi
+
+instance ToMetadata ByteArray where
+  toMetadata = Bytes
+
+instance ToMetadata String where
+  toMetadata = Text
+
+--------------------------------------------------------------------------------
+-- AnyToMetadata
+--------------------------------------------------------------------------------
+
+newtype AnyToMetadata = AnyToMetadata
```

Comment:
My assumption is that this type would be for a heterogeneous collection of `ToMetadata`s, correct (I do think this would be useful for users)? Could we add a doc comment for that?